storage/imxrt-flash/flashsrv.c

```diff
@@ -12,78 +12,290 @@
 }
 
 
-static void flashsrv_devCtl(msg_t *msg)
+static void flashsrv_mfsPartitionErasef1(void)
+{
+	int i;
+	uint32_t sectorsNb;
+	flash_memory_t *flash_memory;
+
+	flash_memory = flashsrv_common.flash_memories + 1;
+	sectorsNb = flash_memory->parts[flash_memory->currPart].sz / flash_memory->ctx.properties.sector_size;
+
+	for (i = 0; i < sectorsNb; ++i)
+		flashsrv_eraseSector(1, flash_memory->parts[flash_memory->currPart].offs + flash_memory->ctx.properties.sector_size * i);
+}
+
+
+static void flashsrv_devCtl(flash_memory_t *memory, msg_t *msg)
 {
-	flash_memory_t *flash_memory = NULL;
 	flash_i_devctl_t *idevctl = (flash_i_devctl_t *)msg->i.raw;
 	flash_o_devctl_t *odevctl = (flash_o_devctl_t *)msg->o.raw;
 
+	if (memory == NULL) {
+		odevctl->err = -EINVAL;
+		return;
+	}
+
 	switch (idevctl->type) {
 		case flashsrv_devctl_properties:
 			TRACE("imxrt-flashsrv: flashsrv_devctl_properties, id: %d, port: %d.", idevctl->oid.id, idevctl->oid.port);
-			flash_memory = flashsrv_findFlash(idevctl->oid);
-			if (flash_memory == NULL) {
-				odevctl->err = -EINVAL;
-				break;
-			}
+			odevctl->properties.size = memory->ctx.properties.size;
+			odevctl->properties.psize = memory->ctx.properties.page_size;
+			odevctl->properties.ssize = memory->ctx.properties.sector_size;
+			odevctl->properties.offs = 0;
+			odevctl->err = EOK;
+			break;
+
+		case flashsrv_devctl_sync:
+			TRACE("imxrt-flashsrv: flashsrv_devctl_sync, id: %d, port: %d.", idevctl->oid.id, idevctl->oid.port);
+			flash_sync(&memory->ctx);
+			odevctl->err = EOK;
+			break;
+
+		case flashsrv_devctl_eraseSector:
+			TRACE("imxrt-flashsrv: flashsrv_devctl_eraseSector - adrr: %d, id: %d, port: %d.", idevctl->erase.addr, idevctl->oid.id, idevctl->oid.port);
+			flashsrv_eraseSector(memory->fOid.id, idevctl->erase.addr);
+			odevctl->err = EOK;
+			break;
 
-			odevctl->properties.fsize = flash_memory->context.properties.size;
-			odevctl->properties.psize = flash_memory->context.properties.page_size;
-			odevctl->properties.ssize = flash_memory->context.properties.sector_size;
+		case flashsrv_devctl_erasePartition:
+			TRACE("imxrt-flashsrv: flashsrv_devctl_eraseChip, id: %d, port: %d.", idevctl->oid.id, idevctl->oid.port);
+			flashsrv_chipErase(memory->fOid.id);
 			odevctl->err = EOK;
 			break;
 
+		default:
+			odevctl->err = -EINVAL;
+			break;
+	}
+}
+
+
+static void flashsrv_rawCtl(flash_memory_t *memory, msg_t *msg)
+{
+	int i;
+	uint32_t sNb;
+	uint8_t partID;
+	flash_i_devctl_t *idevctl = (flash_i_devctl_t *)msg->i.raw;
+	flash_o_devctl_t *odevctl = (flash_o_devctl_t *)msg->o.raw;
+
+	if (memory == NULL) {
+		odevctl->err = -EINVAL;
+		return;
+	}
+
+	switch (idevctl->type) {
+		case flashsrv_devctl_properties:
+			partID = idevctl->oid.id;
+			TRACE("imxrt-flashsrv: flashsrv_devctl_properties, id: %d, port: %d.", partID, idevctl->oid.port);
+			odevctl->properties.size = memory->parts[partID].sz;
```

Comment:
You use data from external source to address the memory without checking it's validity. It's an exploit.